### DISCUSSION TITLE
Ways to make a contract not fail immediately?

### DISCUSSION THREAD

**Discussion by suimong** [tweag/nickel#1828] [Q&A]
Is there a way to make broken contract not fail immediately? e.g. I'm writing a contract that's in effect composed of 4 smaller contracts, and if two of them would fail on a value, I'd like nickel to apply all 4 contracts separately without failing but let me to collect the error info, and when I manually invoke `std.fail_with`  I could let the user know which two contracts passed and which two contracts failed in the error message.

**Comment by yannham** [comment]
Currently there's no way to catch a contract error, if that what's you're referring to. One possibility would be to use predicates instead - that is, have your actual contract code be `foo_predicate = fun value => ...returns some boolean, or maybe some boolean + metadata for label custom errors...`, and just do `Foo = std.contract.from_predicate foo_predicate` (or your own `from_predicate` variant if you want to propagate error message/notes). If the predicate is your original source of truth, then you can easily write a another predicate that checks for `foo_predicate`, `bar_predicate`, etc. - that is any boolean combination.

I'm not sure we'll have some form of general try-catch mechanism for contracts. This has an important impact on the rest of the language. A middle ground would be that there is a better distinction of contracts that are just predicates - for example `Number`, `Bool`, `String`, and any contract build from `from_predicate` - from which we could extract the predicate back with some stdlib helper.

**Reply by suimong** [reply]
Yes catching contract error is what I had in mind. As you mentioned, if the sources of truth come from predicate functions then problem solved. However a big category of contracts, i.e. record contracts, can not be used as predicate AFAIK (should have mentioned this in the original post). Admittedly I prefer the proposed "middleground" than a try-catch mechanism. A helper function that could turn a record contract into predicate would be nice.

**Reply by yannham** [reply]
Indeed. I agree doing any kind of boolean operation on contracts is hard at the moment, even when that should be possible, which is sad. Eventually we would like to have boolean operators, and maybe a predicate extractor (which is, in fact, a very similar problem), that would work on _most_ contracts, and only complain when they can't proceed (function contract, opaque custom contract which is not defined as a predicate, etc.).

**Reply by suimong** [reply]
Is there anything I could do to help? Instinctively I feel like this requires to make a few design decisions, and the implementation might be more intertwined with the internals of Nickel than I am currently comfortable with. Still, if you could provide some guidance I'm happy to give it a shot!

**Reply by vkleen** [reply]
If I remember right, we had the idea of an opaque contract type. That would give us a reasonable way of tracking where a contract comes from (e.g. from a predicate) and then deciding whether boolean operations could make sense. But I don't think we had a good story for how to make this backward compatible. At the moment, any function in Nickel could conceivably be a contract, so there is no good way of tracking a contract through the evaluation of a program.

**Reply by yannham** [reply]
It's not obvious to implement while maintaining backward compatibility, instead. Though a possible middle ground would be to just introduce a new specific object just for contract coming from predicates, and just make them act like functions when applied. It's not the cleanest internal design in the long term but could get us most of the way with boolean stuff while maintaining backward compatibility (plus a long-term detail of not evaluating types away automatically, which is easy and would give us access to builtin contracts)

**Reply by yannham** [reply]
For the record, that's already how match statement works: they're a distinct node in the AST, considered as a value as well, but transparently act as functions when applied.

**Reply by suimong** [reply]
Just wondering, would a record contract count as a "contract coming from predicates"? It does not look particularly "boolish".

**Reply by yannham** [reply]
Yes, you're right. It is boolish if and only if all of the contracts affected to fields are. But sometimes we can also implement boolean operator (at least the `or`, the `not` is more constraining) even if the contracts aren't fully boolish. For example, for `{foo | NonBoolish} \/ {bar | NonBoolish}`, even if `NonBoolish` isn't a predicate, you can derive `has_field "foo"` as a discriminant that tells you directly which branch to take, and thus synthesize the `or` contract as: (I don't use `has_field` but pattern matching for clarity):

```
fun label =>
  match {
    value @ {foo, ..} => std.contract.apply {foo | NonBoolish} label value,
    value @ {bar, ..} => std.contract.apply {bar | NonBoolish} label value,
    _ => std.contract.blame label,
}
```

We discussed that in the meeting, and a special case for `from_predicate` (plus the little adaptation for types) is probably a good first step. Once those internals are implemented, we might think together of an interface for that - happy to ping you then @suimong if help is needed. Implementing at least an `or` on contracts, that is understood by the LSP is long overdue, so let's take the occasion to move that forward a bit.

**Reply by suimong** [reply]
Lovely!
